### PR TITLE
Keep track of the document URL post-redirects

### PR DESCRIPTION
**Pull request by paulirish**
Often from a non-extension the input URL may then be redirected. For example, there are 3 redirects if you attempt `http://aliexpress.com`: 

image

There's a few places where we work with the current URL, including @jeffposnick's new #578 and the [unregsw](https://github.com/GoogleChrome/lighthouse/tree/unregsw) branch. And in these cases, we want the post-redirect URL.

This PR will update `options.url` as redirects change the document location.

Additionally, we used to have a `driver.url`, but only for the chrome extension. We now drop that completely, which simplifies our URL gatherer.

### CODE REVIEW

File: lighthouse-core/gather/drivers/extension.js [hunk 1]
```diff
@@ -60,8 +59,8 @@ class ExtensionDriver extends Driver {
     }
 
     return this.queryCurrentTab_()
-      .then(tabId => {
-        this._tabId = tabId;
+      .then(tab => {
+        const tabId = this._tabId = tab.id;
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/extension.js]: what's the purpose of this change?

Comment (brendankenny) [inline lighthouse-core/gather/drivers/extension.js]: oh I see, jk

File: lighthouse-core/gather/gatherers/url.js [hunk 2]
```diff
@@ -21,7 +21,11 @@ const Gatherer = require('./gatherer');
 class URL extends Gatherer {
 
   beforePass(options) {
-    this.artifact = options.url || options.driver.url;
+    // Used currently by cache-start-url audit, which wants to know if the start_url
+    // in the manifest is stored in the cache.
+    // Instead of the originally inputted URL (options.initialUrl), we want the resolved
+    // post-redirect URL (which is here at options.url)
+    this.artifact = options.url;
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/url.js]: wouldn't you want this `afterPass`, otherwise redirection wouldn't have happened yet?

File: lighthouse-core/gather/gather-runner.js [hunk 3]
```diff
@@ -197,6 +200,11 @@ class GatherRunner {
                 // Merge pass trace and network data into tracingData.
                 config.trace && Object.assign(tracingData.traces, loadData.traces);
                 config.network && (tracingData.networkRecords = loadData.networkRecords);
+              })
+              .then(_ => {
+                // Update the outer options object of any post-redirect url
+                // Subsequent passes will use the new URL, and it'll be reported outwards
+                options.url = runOptions.url;
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: seems like this could bite us, though. Someone adding a new first pass, moving the performance gatherers to the second pass, would miss the redirects in the load time since the second pass would skip to the new URL. Is there a reason each pass shouldn't start with the same URL each time?

File: lighthouse-core/gather/gatherers/url.js (hunk 2 above)
Comment (paulirish) [inline lighthouse-core/gather/gatherers/url.js]: yeah definitely! even better.

File: lighthouse-core/gather/gather-runner.js (hunk 3 above)
Comment (paulirish) [inline lighthouse-core/gather/gather-runner.js]: Did it mostly to save time in our later passes, but I agree it changes behavior. I'm fine with reusing the same initialUrl for each pass

Comment (paulirish) [inline lighthouse-core/gather/gather-runner.js]: Interesting challenge here. Thee http-redirect gather changes the `options.url` in beforePass and resets in afterPass.

They are separated into separate passes, so we are safe for now, but we need to make sure they don't share a pass. (well, until we fix the http redirect gather)

**Comment by paulirish** [comment]
Rebased to master.  

To test, run `lighthouse http://aliexpress.com` and verify output includes: 
`Lighthouse results: https://m.aliexpress.com/?tracelog=wwwhome2mobilesitehome`

File: lighthouse-core/gather/gatherers/http-redirect.js [hunk 4]
```diff
@@ -18,6 +18,11 @@
 
 const Gatherer = require('./gatherer');
 
+/**
+ * This gatherer changes the options.url so that its pass loads the http page.
+ * After load it detects if its on a crypographic scheme.
+ * TODO: Instead of abusing a loadPage pass for this test, it could likely just do an XHR instead
+ */
 class HTTPRedirect extends Gatherer {
```
Comment (paullewis) [inline lighthouse-core/gather/gatherers/http-redirect.js]: Actually wondering if we should be surfacing this. For example, redirecting to an m. incurs a perf penalty, which should be captured in TTI etc, but as a meta piece of data highlighting the number of redirects might be helpful.

**Comment by paullewis** [comment]
LGTM. Think we might want to surface the number of redirects as a piece of data somewhere, but that can be a separate PR.

File: lighthouse-core/gather/drivers/driver.js [hunk 5]
```diff
@@ -219,6 +210,28 @@ class Driver {
   }
 
   /**
+   * If our main document URL redirects, we will update options.url accordingly
+   * As such, options.url will always represent the post-redirected URL.
+   * options.initialUrl is the pre-redirect URL that things started with
+   *
+   * Caveat: only works when network recording enabled for a pass
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: should we file a bug for this? I'm thinking either network recording should always happen, getting rid of `config.network`, or network recording should always happen but the results are discarded if `config.network` is `false`

Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: seems important for gatherers/audits to be able to assume they don't have to do any work to get the actual url of the current page

File: lighthouse-core/gather/drivers/driver.js [hunk 6]
```diff
@@ -219,6 +210,28 @@ class Driver {
...
+   */
+  updateUrlIfRedirected(opts) {
+    const networkManager = this._networkRecorder.networkManager;
+    networkManager.addEventListener(this._networkRecorder.EventTypes.RequestFinished, request => {
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: should add an event listener interface on `NetworkRecorder` so we don't have to reach into it here?

File: lighthouse-core/gather/drivers/driver.js [hunk 7]
```diff
@@ -328,10 +341,11 @@ class Driver {
     });
   }
 
-  beginNetworkCollect() {
+  beginNetworkCollect(opts) {
     return new Promise((resolve, reject) => {
       this._networkRecords = [];
       this._networkRecorder = new NetworkRecorder(this._networkRecords);
+      this.updateUrlIfRedirected(opts);
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: total nit: maybe something like `enableUrlUpdateIfRedirected` to make it clear it's just turning on the listener?

File: lighthouse-core/gather/drivers/extension.js [hunk 8]
```diff
@@ -142,18 +141,16 @@ class ExtensionDriver extends Driver {
           return reject(chrome.runtime.lastError);
         }
 
-        this.url = tabs[0].url;
-        resolve(tabs[0].id);
+        resolve(tabs[0]);
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/extension.js]: while you're in here, should this be caching the tab found and always resolving on it in subsequent calls? Seems bad that a call to `getCurrentTabURL()` could potentially return a different tab than when `connect()` is finally called

Comment (brendankenny) [inline lighthouse-core/gather/drivers/extension.js]: could be storing `this._tab` instead of `this._tabId` and just accessing the `id` prop when needed elsewhere

File: lighthouse-core/gather/gatherers/http-redirect.js (hunk 4 above)
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/http-redirect.js]: redirects do show up (but aren't especially highlighted) in the critical request chains though timing info isn't included

File: lighthouse-core/gather/gather-runner.js [hunk 9]
```diff
@@ -196,18 +196,30 @@ class GatherRunner {
...
+              })
+              .then(_ => {
+                if (passIndex === 0) {
+                  urlAfterRedirects = runOptions.url;
+                }
               });
-        }, Promise.resolve());
+        }, Promise.resolve()).then(_ => {
+          options.url = urlAfterRedirects;
+        });
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: since you no longer alter the options object in this loop you can simplify this again:
```js
.then(_ => {
  // If the main document redirects, we'll update this to keep track
  let urlAfterRedirects;
  return passes.reduce((chain, config, passIndex) => {
    const runOptions = Object.assign({}, options, {config});
    return chain
      .then(_ => GatherRunner.beforePass(runOptions))
      .then(_ => GatherRunner.pass(runOptions))
      .then(_ => GatherRunner.afterPass(runOptions))
      .then(loadData => {
        // Merge pass trace and network data into tracingData.
        config.trace && Object.assign(tracingData.traces, loadData.traces);
        config.network && (tracingData.networkRecords = loadData.networkRecords);
        if (passIndex === 0) {
          urlAfterRedirects = runOptions.url;
        }
      });
  }, Promise.resolve()).then(_ => {
    options.url = urlAfterRedirects;
  });
})
```
(technically you could even set `options.url` to `runOptions.url` in that `passIndex` conditional since each pass already has its own copy of options but that might be confusing)

Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: `passIndex === 0` is going to break in configs with `config.network === false` in the first pass

**Comment by brendankenny** [comment]
It still feels a bit weird to be updating `options.url` as we go. One other option: it could be collected as part of the tracing data (along with the trace and network records), passed into `afterPass()` on the `loadData` object and merged into the artifacts along with them as well

**Comment by brendankenny** [comment]
(we could explore that in a follow up with a general `options` cleanup)

**Comment by paulirish** [comment]
Yup I like the idea of doing it alongside the trace and network data.

File: lighthouse-core/gather/drivers/extension.js (hunk 8 above)
Comment (paulirish) [inline lighthouse-core/gather/drivers/extension.js]: > Seems bad that a call to getCurrentTabURL() could potentially return a different tab than when connect() is finally called

Thinking about it, I don't think it's a risk. (Although I admit it's a bit odd.)
First `getCurrentTabURL` is called via extension popup to get the URL to do the run with. Later, `connect()` gets the activeTab in order to figure out which tab to control.  But this happens v quickly and even if the user manages to switch to a diff active tab within those 50ms, I think that's okay.

File: lighthouse-core/gather/gather-runner.js (hunk 9 above)
Comment (paulirish) [inline lighthouse-core/gather/gather-runner.js]: > passIndex === 0 is going to break in custom configs with config.network === false in the first pass, though

K one more reason to do #604 ;)

File: lighthouse-core/gather/drivers/driver.js (hunk 6 above)
Comment (paulirish) [inline lighthouse-core/gather/drivers/driver.js]: sg.

**Comment by paulirish** [comment]
A few items came up from this. Filed in #604 and #605. 
Otherwise we should be good to go.

File: lighthouse-core/gather/drivers/driver.js (hunk 5 above)
Comment (paulirish) [inline lighthouse-core/gather/drivers/driver.js]: filed #604
